src/controllers/v1/certificate/__tests__/index.test.ts

```diff
@@ -0,0 +1,646 @@
+}
+
+const transactionExample = {
+  id: 'initiate-cert-transaction-test',
+  api_type: 'certificate',
+  local_id: 'test-cert-1',
+  state: 'submitted',
+  transaction_type: 'initiate_cert',
+}
+const certExamples = [
+  { id: 'test-cert-1', state: 'issued', energy_owner: 'emma-test', regulator: 'ray-test', hydrogen_owner: 'heidi' },
+  { id: 'test-cert-4', state: 'initiated', energy_owner: 'emma2-test', regulator: 'ray-test', hydrogen_owner: 'heidi' },
+  {
+    id: 'test-cert-3',
+    state: 'initiated',
+    energy_owner: 'member-self-test',
+    regulator: 'ray-test',
+    latest_token_id: 3,
+    hydrogen_owner: 'member-self-test',
+    hydrogen_quantity_mwh: 10,
+    commitment: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
+    production_start_time: new Date('2024-01-03T15:17:58.836Z'),
+    production_end_time: new Date('2024-01-03T18:17:58.836Z'),
+    energy_consumed_mwh: 5,
+  },
+  {
+    id: 'test-cert-2',
+    state: 'pending',
+    energy_owner: 'member-self-test',
+    regulator: 'ray-test',
+    hydrogen_owner: 'member-self-test',
+    hydrogen_quantity_mwh: 10,
+  },
+  {
+    id: 'test-cert-5',
+    state: 'initiated',
+    energy_owner: 'emma-test',
+    regulator: 'member-self-test',
+    hydrogen_owner: 'heidi',
+    hydrogen_quantity_mwh: 10,
+    commitment: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
+    production_start_time: new Date('2024-01-03T15:17:58.836Z'),
+    production_end_time: new Date('2024-01-03T18:17:58.836Z'),
+    energy_consumed_mwh: 5,
+  },
+]
+
+describe('v1/certificate', () => {
+  let response: any
+  let controller: CertificateController
+  const database: Database = new Database()
+  const identity: Identity = new Identity(new Env())
+  const commitment: Commitment = new Commitment('sha256')
+  const node: ChainNode = new ChainNode(new Env(), database)
+  const emissions = new EmissionsCalculator()
+
+  const stubs = {
+    build: sinon.stub(commitment, 'build' as any).callsFake(() => 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'),
+    getSelfAddress: sinon
+      .stub(identity, 'getMemberBySelf' as any)
+      .callsFake((alias) => ({ address: 'member-self-test', alias })),
+    getMemberByAlias: sinon
+      .stub(identity, 'getMemberByAlias' as any)
+      .callsFake((alias) => ({ address: 'member-by-alias-test', alias })),
+    prepareRunProcess: sinon
+      .stub(node, 'prepareRunProcess' as any)
+      .callsFake(() =>
+        Promise.resolve({ hash: { toHex: () => '9d441a0fe4fb942070f4d3014e2367496d4afc3bc9b983f1ac5b3813467a0c19' } })
+      ),
+    submitRunProcess: sinon.stub(node, 'submitRunProcess' as any).callsFake(() => Promise.resolve()),
+    get: sinon.stub(database, 'get').resolves(certExamples as CertificateRow[]),
+    insert: sinon.stub(database, 'insert' as any).callsFake((_, data) => [data]),
+    getLastFinalisedBlockHash: sinon
+      .stub(node, 'getLastFinalisedBlockHash')
+      .callsFake(() => '0x9d441a0fe4fb942070f4d3014e2367496d4afc3bc9b983f1ac5b3813467a0c19' as any),
+  }
```

Comment:
This feels a little fragile. My main concern is that these stubs are not being set for each test except where you're calling `resetHistory` and even that isn't a complete reset. My worry is one failure contaminates further tests because the stub is shared but has been modified. For example:

```js
describe('if certificate does not exist', () => {
      beforeEach(async () => {
        stubs.get.throws(new NotFound('certificate'))
        response = await controller.getById('does-nt-exists-id').catch((err) => err)
      })
     ...
```

Mutates the stub in a way that might affect other tests. Perhaps a better way would be to generate the `stubs` object in a helper function that can be called in a `beforeEach`? That can then set the stubs on a context object that's passed in but the function can make sure all stubs are reset correctly. 

Thoughts?

Reply:
Fix in future PR as this isn't a regression